include/wil/win32_helpers.h

```diff
@@ -286,6 +286,56 @@ namespace wil
     }
 #pragma endregion
 
+#pragma region RECT helpers
+    template<typename rect_type>
+    constexpr auto rect_width(rect_type rect)
+    {
+        return rect.right - rect.left;
+    }
+
+    template<typename rect_type>
+    constexpr auto rect_width(rect_type* rect)
+    {
+        return rect->right - rect->left;
+    }
+
+    template<typename rect_type>
+    constexpr auto rect_height(rect_type rect)
+    {
+        return rect.bottom - rect.top;
+    }
+
+    template<typename rect_type>
+    constexpr auto rect_height(rect_type* rect)
+    {
+        return rect->bottom - rect->top;
+    }
+
+    template<typename rect_type>
+    constexpr auto rect_is_empty(rect_type rect)
+    {
+        return (rect.left >= rect.right) || (rect.top >= rect.bottom);
+    }
+
+    template<typename rect_type>
+    constexpr auto rect_is_empty(rect_type* rect)
+    {
+        return (rect->left >= rect->right) || (rect->top >= rect->bottom);

```

Comment:
It is not ideal that the pointer version duplicates the value type version for each of these methods.  Would it make more sense to call across instead?

```suggestion
        return rect_is_empty(*rect);
```